### PR TITLE
MPLS EXP to TC mapping HLD

### PR DESCRIPTION
**Pull request by abanu-ms**
This HLD presents the design for adding the MPLS EXP to TC QoS mapping, along the existing DSCP to TC mapping.

Code PRs:
swss-common: https://github.com/Azure/sonic-swss-common/pull/537
swss: https://github.com/Azure/sonic-swss/pull/1954
utilites: https://github.com/Azure/sonic-utilities/pull/1875

### CODE REVIEW

File: doc/qos/mpls_exp_to_tc_map.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+TC = Traffic Class
+QoS = Quality of Service
+
+## 5. Overview
+
+This new enhancement adds support to SONiC for MPLS EXP to TC map which allows QoS to work on MPLS packets.
+
+## 6. Requirements
+
+User can configure MPLS EXP to TC map at start-of-day via configuration file. CLI support will exist to offer the same amount of support as for DSCP to TC map.
```
Comment (venkatmahalingam) [inline doc/qos/mpls_exp_to_tc_map.md:51]: MPLS EXP has been renamed as MPLS TC as per rfc5462.

File: doc/qos/mpls_exp_to_tc_map.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+## 7. Architecture Design
+
+The overall SONiC architecture will not be changed and no new sub-modules will be introduced.
+
+## 8. High-Level Design
+
+### 8.1. DB
+
+The CONFIG DB will be updated to include a new "MPLS_EXP_TO_TC_MAP_TABLE" similar to the existing "DSCP_TO_TC_MAP_TABLE". This will have the following format:
```
Comment (venkatmahalingam) [inline doc/qos/mpls_exp_to_tc_map.md:61]: We're changing the DSCP_TO_TC_MAP_TABLE table representation in the config-DB now, please refer to the below PR and ensure to follow the same model for MPLS TC to TC MAP table as well.
https://github.com/Azure/sonic-buildimage/pull/7375/files

Comment (abanu-ms) [reply doc/qos/mpls_exp_to_tc_map.md]: Is there any expected date when that PR is going to get in? We'd want the MPSL TC to TC MAP PR to get in for the November release. If the referenced PR isn't going to be merged soon or at least the existing comments to be addressed in order to come to a conclusion about how the map/list structure should look like, we could proceed with the existing design and come back and update it when that PR is also merged.

Comment (abanu-ms) [reply doc/qos/mpls_exp_to_tc_map.md]: Following up on the today's meeting for clarity - is the existing `DSCP_TO_TC_MAP` table structure an issue? If not, the new table is going to look exactly the same for MPLS. If it is an issue, could you please explain how you think the table structure should be updated?

**Review by rck-innovium — CHANGES_REQUESTED** [review]
A user can bind the MPLS_TC_TO_TC_MAP to the switch, port, and/or MPLS in-segment entry using the below SAI attributes:
 SAI_SWITCH_ATTR_QOS_MPLS_EXP_TO_TC_MAP or SAI_PORT_ATTR_QOS_MPLS_EXP_TO_TC_MAP or  SAI_INSEG_ENTRY_ATTR_MPLS_EXP_TO_TC_MAP.


The proposal defines an MPLS_TC_TO_TC_MAP.  However it does not specify how an user can specify the binding of the map.

**Comment by abanu-ms** [comment]
> A user can bind the MPLS_TC_TO_TC_MAP to the switch, port, and/or MPLS in-segment entry using the below SAI attributes:
> SAI_SWITCH_ATTR_QOS_MPLS_EXP_TO_TC_MAP or SAI_PORT_ATTR_QOS_MPLS_EXP_TO_TC_MAP or SAI_INSEG_ENTRY_ATTR_MPLS_EXP_TO_TC_MAP.
> 
> The proposal defines an MPLS_TC_TO_TC_MAP. However it does not specify how an user can specify the binding of the map.

I've updated the proposal to include per-port binding configuration details.

At this point we're not going to offer support for per-switch and per-inseg configuration as the scope is to offer the same amount of support as DSCP to TC map does in the beginning.

**Comment by zhangyanzhao** [comment]
@abanu-ms  can you please add the code PRs into this HLD PR by referring to the example https://github.com/Azure/SONiC/pull/806

File: doc/qos/mpls_exp_to_tc_map.md (hunk 2 above)
Comment (TACappleman) [reply doc/qos/mpls_exp_to_tc_map.md]: Our code now follows the new model

**Comment by TACappleman** [comment]
@zhangyanzhao can this doc be merged? (All the code PRs for it are now in)

**Pull request merged**